src/displayBackend/drm/DrmDeviceDetector.cpp

```diff
@@ -0,0 +1,162 @@
+ *   along with this program; if not, write to the Free Software
+ *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
+ *
+ * Copyright (C) 2020 EPAM Systems Inc.
+ *
+ */
+#include "DrmDeviceDetector.hpp"
+
+#include <memory>
+#include <string>
+#include <stdexcept>
+
+#include <libudev.h>
+#include <xf86drm.h>
+#include <xf86drmMode.h>
+#include <fcntl.h>
+
+#include <xen/be/Log.hpp>
+
+namespace Drm {
+
+struct drmFdGuard {
+public:
+
+	drmFdGuard(int fd) : mFd(fd) {}
+
+	~drmFdGuard()
+	{
+		if (mFd > 0)
+		{
+			drmClose(mFd);
+		}
+	}
+
+	operator int() const { return mFd; }
+
+	drmFdGuard(const drmFdGuard&) = delete;
+	drmFdGuard& operator=(drmFdGuard const&) = delete;
+
+private:
+
+	int mFd;
+};
+
+static bool isDrmDevice(udev_device* device)
+{
+	auto fileName = udev_device_get_devnode(device);
+	auto sysNum = udev_device_get_sysnum(device);
+
+	if (!fileName || !sysNum || atoi(sysNum) < 0)
+	{
+		return false;
+	}
+
+	drmFdGuard fd(open(fileName, O_RDWR | O_CLOEXEC));
+	if (fd < 0)
+	{
+		return false;
+	}
+
+	std::unique_ptr<drmModeRes, decltype(&drmModeFreeResources)>
+			res(drmModeGetResources(fd), drmModeFreeResources);
+	if (!res)
+	{
+		return false;
+	}
+
+	if (res->count_crtcs <= 0 || res->count_connectors <= 0 ||
+	    res->count_encoders <= 0)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+struct udevError : std::runtime_error
+{
+	udevError(const char*msg) : std::runtime_error(msg) {};
+};
+
+static void checkFail(bool fail, const char* msg)
+{
+	if (fail)
+	{
+		throw udevError(msg);
+	}
+}
+
+std::string detectDrmDevice()
+{
+	XenBackend::Log log("DrmDeviceDetector");
+
+	LOG(log, INFO) << "Auto detecting DRM KMS device";
+	try {
+		std::unique_ptr<struct udev, decltype(&udev_unref)>
+				udev(udev_new(), &udev_unref);
+
+		checkFail(!udev, "Cannot create udev context");
+
+		std::unique_ptr<udev_enumerate, decltype(&udev_enumerate_unref)>
+				e(udev_enumerate_new(udev.get()), udev_enumerate_unref);
+		checkFail(!e, "Cannot create udev enumerator");
+
+		checkFail(udev_enumerate_add_match_subsystem(e.get(), "drm"),
+				   "Error adding subsystem match");
+		checkFail(udev_enumerate_add_match_sysname(e.get(), "card[0-9]*"),
+				   "Error adding sysname match");
+		checkFail(udev_enumerate_scan_devices(e.get()),
+				   "Error scanning for udev devices");
+
+		udev_list_entry *entry;
+		udev_list_entry_foreach(entry, udev_enumerate_get_list_entry(e.get()))
+		{
+			auto path = udev_list_entry_get_name(entry);
+			if (!path)
+			{
+				continue;
+			}
+
+			std::unique_ptr<struct udev_device, decltype(&udev_device_unref)>
+					device(udev_device_new_from_syspath(udev.get(), path),
+						   udev_device_unref);
+			if (!device)
+			{
+				continue;
+			}
+
+			if (isDrmDevice(device.get()))
+			{
+				auto fileName = udev_device_get_devnode(device.get());
+				LOG(log, INFO) << "Using " << fileName;
+
+				return fileName;
+			}
+		}
+	}
+	catch(udevError &err)
```

Comment:
Why we need catch here? General in project there are only one error handler per thread. I guess we can just throw exception in this function excepts  situation when "" valid case.

Reply:
This is a convience wrapper. Before this change, the code looked like this:
https://github.com/lorc/libxenbe/commit/6704584c01f69ff8e72226b86accae1fa49fe09b
I added this fix to reduce amount of boilerplate.
https://github.com/lorc/libxenbe/commit/7f998f248c50232aa69bbba387a9cdcbe9cf4a20

Reply:
So yes, actually, we don't want to propagate udev errors to the caller. This is our problem and we handle it by telling "sorry, can't find the default DRM device" to the caller.

Reply:
it is called in wayland context, so we need to use try-catch here

Reply:
anyways, it leads to an exception on Wayland's dispatch thread which currently kills the backend with SIGTERM....